launcher/ui/themes/CatPack.cpp

```diff
@@ -0,0 +1,111 @@
+ */
+
+#include "ui/themes/CatPack.h"
+#include <qdatetime.h>
+#include <qjsonarray.h>
+#include <qjsonobject.h>
+#include <qobject.h>
+#include <QDateTime>
+#include <QDir>
+#include <QFileInfo>
+#include "FileSystem.h"
+#include "Json.h"
+#include "ui/themes/ThemeManager.h"
+
+QString BasicCatPack::path()
+{
+    const QDateTime now = QDateTime::currentDateTime();
+    const QDateTime birthday(QDate(now.date().year(), 11, 30), QTime(0, 0));
+    const QDateTime xmas(QDate(now.date().year(), 12, 25), QTime(0, 0));
+    const QDateTime halloween(QDate(now.date().year(), 10, 31), QTime(0, 0));
+
+    QString cat = QString(":/backgrounds/%1").arg(m_id);
+    if (std::abs(now.daysTo(xmas)) <= 4) {
+        cat += "-xmas";
+    } else if (std::abs(now.daysTo(halloween)) <= 4) {
+        cat += "-spooky";
+    } else if (std::abs(now.daysTo(birthday)) <= 12) {
+        cat += "-bday";
+    }
+    return cat;
+}
+
+JsonCatPack::JsonCatPack(QFileInfo& manifestInfo) : BasicCatPack(manifestInfo.dir().dirName())
+{
+    QString path = FS::PathCombine("catpacks", m_id);
+
+    if (!FS::ensureFolderPathExists(path)) {
+        themeWarningLog() << "couldn't create folder for catpack!";
+        return;
+    }
+
+    if (manifestInfo.exists() && manifestInfo.isFile()) {
+        try {
+            auto doc = Json::requireDocument(manifestInfo.absoluteFilePath(), "CatPack JSON file");
+            const auto root = doc.object();
+            m_name = Json::requireString(root, "name", "Catpack name");
+            auto id = Json::ensureString(root, "id", "", "Catpack ID");
+            if (!id.isEmpty())
+                m_id = id;
+            m_defaultPath = FS::PathCombine(path, Json::requireString(root, "default", "Deafult Cat"));
+            auto variants = Json::ensureArray(root, "variants", QJsonArray(), "Catpack Variants");
+            for (auto v : variants) {
+                auto variant = Json::ensureObject(v, QJsonObject(), "Cat variant");
+                m_variants << Variant{ FS::PathCombine(path, Json::requireString(variant, "path", "Variant path")),
+                                       date(Json::requireString(variant, "startTime", "Variant startTime")),
+                                       date(Json::requireString(variant, "endTime", "Variant endTime")) };
+            }
+
+        } catch (const Exception& e) {
+            themeWarningLog() << "Couldn't load catpack json: " << e.cause();
+            return;
+        }
+    } else {
+        themeDebugLog() << "No catpack json present.";
+    }
+}
+
+QString JsonCatPack::path()
+{
+    const QDateTime now = QDateTime::currentDateTime();
+    for (auto var : m_variants) {
+        QDateTime startDate(QDate(now.date().year(), var.startTime.mounth, var.startTime.day), QTime(0, 0));
+        QDateTime endDate(QDate(now.date().year(), var.endTime.mounth, var.endTime.day), QTime(0, 0));
+        if (startDate.daysTo(now) > 0 && now.daysTo(endDate) > 0)
+            return var.path;
```

Comment:
could be nice to support having the end date before the start date to wrap across year a boundary (although i guess you could just set two variants with the same image)

also in my implementation i just compared `QDate`'s, does `QDateTime` provide some advantage here?

Reply:
I added the wrap: so now if in the current year start date is after the end date I added a year to end date.
Also regarding QDateTime vs QDate, there was no reason behind, now it is QDate